Docs updates for Bitbucket Server SCM Flow (#500)
algorithmiaio/dev-center: _pages/algorithm-development/source-control-management.md

### Hosting Source Code on Bitbucket Cloud

By hosting your algorithm's source code on Bitbucket Cloud, you can take advantage of Bitbucket Cloud's rich set of developer features, such as pull requests, and also ensure that access to your source code is carefully mediated.

If you're new to Git, we recommend [this tutorial series](https://try.github.io/) by GitHub.

**Web IDE Support** At this time we do not support editing source code in our web app for Bitbucket Cloud-hosted algorithms. 
{: .notice-info}

#### Connecting Your Algorithmia & Bitbucket Cloud Accounts

To connect your Algorithmia and Bitbucket Cloud accounts, simply select the Bitbucket Cloud instance you wish to use when creating your algorithm:

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_bitbucket_algorithm_unauthorized.png" alt="Authorization prompt for Bitbucket Cloud repository host" class="screenshot img-sm">

> If you haven't yet connected your Bitbucket Cloud account, you will be prompted to do so. By connecting your Bitbucket Cloud account, you will give us access to create repositories on your behalf, which will house your algorithm's source code.

Reviewer: for a few grammar issues, change to:
If you haven't yet connected your Bitbucket Cloud account, you'll be prompted to do so. Your Bitbucket Cloud account will house your algorithms' source code, and by connecting it, you'll give us access to create repositories on your behalf.

> If you haven't yet connected your Bitbucket Cloud account, you will be prompted to do so. By connecting your Bitbucket Cloud account, you will give us access to create repositories on your behalf, which will house your algorithm's source code.

Reviewer: ```suggestion
If you haven't yet connected your Bitbucket Cloud account, you'll be prompted to do so. Your Bitbucket Cloud account will house your algorithms' source code, and by connecting it, you'll give us access to create repositories on your behalf.
```

Reviewer: this fixes a few grammar issues

#### Creating Algorithms with Bitbucket Cloud

Once you've connected your Bitbucket Cloud account to Algorithmia, you will be able to host algorithm source code in Bitbucket Cloud.

> You can customize two aspects of the Bitbucket Cloud repositories that are created for your algorithms: the repository's owner, and its name.

Reviewer: plurality conflict, change to:
You can customize two aspects of any Bitbucket Cloud repository that's created for an algorithm: ...

> You can customize two aspects of the Bitbucket Cloud repositories that are created for your algorithms: the repository's owner, and its name.

Reviewer: ```suggestion
You can customize two aspects of any Bitbucket Cloud repository that's created for an algorithm: the repository's owner, and its name.
```

> By default, all Bitbucket Cloud respositories are created under your personal Bitbucket Cloud account. However, you may optionally choose any Bitbucket Cloud organizations that you are a member of. These organizations are listed in the "Repository Owner" dropdown:

Reviewer: correct typo "respositories" ;

suggest change to:
However, you may optionally choose to create repositories under any Bitbucket Cloud organizations of which you're a member.

> By default, all Bitbucket Cloud respositories are created under your personal Bitbucket Cloud account. However, you may optionally choose any Bitbucket Cloud organizations that you are a member of. These organizations are listed in the "Repository Owner" dropdown:

Reviewer: ```suggestion
By default, all Bitbucket Cloud repositories are created under your personal Bitbucket Cloud account. However, you may optionally choose to create repositories under any Bitbucket Cloud organizations of which you're a member. These organizations are listed in the "Repository Owner" dropdown:
```

__I revoked access to the Algorithmia's GitLab OAuth application, and now none of the repositories I created can build!__

When you revoke access to an OAuth app, GitLab automatically revokes both your token and any deploy keys you may have created via the OAuth app. To fix your algorithms, simply [follow the instructions above](#your-repositorys-deploy-key-was-removed) on restoring deploy keys to a repository.

### Hosting Source Code on Bitbucket Server

> **No Oauth Support** Bitbucket Server does not support OAuth, so a token based flow is implemented to connect Bitbucket Server-hosted algorithms. 

Reviewer: --> token-based [add hyphen]

> **No Oauth Support** Bitbucket Server does not support OAuth, so a token based flow is implemented to connect Bitbucket Server-hosted algorithms. 

Reviewer: ```suggestion
**No Oauth Support** Bitbucket Server does not support OAuth, so a token-based flow is implemented to connect Bitbucket Server-hosted algorithms. 
```

{: .notice-info}

By hosting your algorithm's source code on Bitbucket Server, you can take advantage of Bitbucket Server's rich set of developer features, such as pull requests, and also ensure that access to your source code is carefully mediated.

If you're new to Git, we recommend [this tutorial series](https://try.github.io/) by GitHub.

> **Web IDE Support** At this time we do not support editing source code in our web app for Bitbucket Server-hosted algorithms. 

Reviewer: web app --> Web IDE

> **Web IDE Support** At this time we do not support editing source code in our web app for Bitbucket Server-hosted algorithms. 

Reviewer: ```suggestion
**Web IDE Support** At this time we do not support editing source code in our Web IDE for Bitbucket Server-hosted algorithms. 
```

{: .notice-info}

#### Creating Algorithms with Bitbucket Server

To host your algorithm on Bitbucket Server, simply select the Bitbucket Server instance you wish to use when creating your algorithm.

Once you've selected your Bitbucket Server instance, you will be able to host algorithm source code in Bitbucket Server.

> You can customize two aspects of the Bitbucket Server repositories that are created for your algorithms: the repository's owner, and its name.

Reviewer: change to:
You can customize two aspects of any Bitbucket Server repository that's created for an algorithm: ...

> You can customize two aspects of the Bitbucket Server repositories that are created for your algorithms: the repository's owner, and its name.

Reviewer: ```suggestion
You can customize two aspects of any Bitbucket Server repository that's created for an algorithm: the repository's owner, and its name.
```

> By default, all Bitbucket Server respositories are the project configured by your administrator. These proojects are listed in the "Repository Owner" dropdown:

Reviewer: typo "respositories" ;
First sentence meaning is unclear.

> By default, all Bitbucket Server respositories are the project configured by your administrator. These proojects are listed in the "Repository Owner" dropdown:

Reviewer: ```suggestion
By default, all Bitbucket Server repositories are the project configured by your administrator. These projects are listed in the "Repository Owner" dropdown:
```

Reviewer: Also, first sentence meaning is unclear.

> You may also customize the name of the repository that is created. By default the name of your algorithm is used. 

Reviewer: This is redundant with what's listed two paragraphs back and should be combined with that.

> 

Reviewer: they will replace --> it will replace

> If you decide to use an alternative name, we recommend that you use only letters, numbers, hyphens, and/or underscores. While Bitbucket Server will accept characters outside this range, they will replace any unsupported characters with an underscore, resulting in an undesirable repository name.

Reviewer: they will replace --> it will replace

> If you decide to use an alternative name, we recommend that you use only letters, numbers, hyphens, and/or underscores. While Bitbucket Server will accept characters outside this range, they will replace any unsupported characters with an underscore, resulting in an undesirable repository name.

Reviewer: ```suggestion
If you decide to use an alternative name, we recommend that you use only letters, numbers, hyphens, and/or underscores. While Bitbucket Server will accept characters outside this range, it will replace any unsupported characters with an underscore, resulting in an undesirable repository name.
```

When we create your repository, we associate the following with it:

> - **A Personal Access Token**: Bitbucket Server personal access tokens allow appropriate access to specific repositories, and are Bitbucket Server’s prescribed means by which external services can fetch code for building and deploying. These tokens are tied to individual permissions, and as such will not allow Algorithmia to continue building an algorithm if the permissions of the Personal Access Token changes. We also use the Personal Access Token to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).

Reviewer: lowercase "personal access token" in body

> - **A Personal Access Token**: Bitbucket Server personal access tokens allow appropriate access to specific repositories, and are Bitbucket Server’s prescribed means by which external services can fetch code for building and deploying. These tokens are tied to individual permissions, and as such will not allow Algorithmia to continue building an algorithm if the permissions of the Personal Access Token changes. We also use the Personal Access Token to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).

Reviewer: ```suggestion
- **A Personal Access Token**: Bitbucket Server personal access tokens allow appropriate access to specific repositories, and are Bitbucket Server’s prescribed means by which external services can fetch code for building and deploying. These tokens are tied to individual permissions, and as such will not allow Algorithmia to continue building an algorithm if the permissions of the personal access token changes. We also use the personal access token to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).
```

- **Webhooks**: We set up webhooks to receive notifications about changes to your repo, such as when there's a change to its default branch.

#### Updating Bitbucket Server-Hosted Algorithms

Once you've created your algorithm, any commits to your Bitbucket Server repository's _default_ branch will result in a build on Algorithmia. At creation, your repository's default branch will be your `master` branch, but you can change this at any time in your repository's settings.

> You can view your algorithm's builds by heading to its landing page and clicking the "Builds" tab if you are the algorithm's owner.

Reviewer: landing page --> homepage

> You can view your algorithm's builds by heading to its landing page and clicking the "Builds" tab if you are the algorithm's owner.

Reviewer: ```suggestion
You can view your algorithm's builds by heading to its homepage and clicking the "Builds" tab if you are the algorithm's owner.
```

#### Publishing Bitbucket Server-Hosted Algorithms

> When you're ready to create a new version of your algorithm, visit that algorithm's page and select the __Builds__ tab. Locate the build corresponding to specific commit SHA you wish to use, and then click __Publish__ on the right side of the row.

Reviewer: Formatting of "Builds" from previous paragraph and __Builds__ from this paragraph is different.

This will open a publish modal that will let you publish your algorithm.

If you don't see any builds listed, make sure that you're pushing commits to your repository's _default_ branch. Builds are only triggered for commits pushed to the default branch.

To learn more about algorithm builds, [click here]({{site.cdnurl}}{{site.baseurl}}/platform/algorithm-profiles#algorithm-builds).

#### Deleting Bitbucket Server-Hosted Algorithms

If you wish to delete an algorithm that hosts its source code in Bitbucket Server, have no fear: the repository will be left unharmed. 

#### Managing Your Bitbucket Server Authorization

> If you want to review your Bitbucket Server authorization status, you can visit your user settings page. Simply scroll to the "Source Control Management" section to view any prior Bitbucket Server authorizations

Reviewer: add period
```suggestion
If you want to review your Bitbucket Server authorization status, you can visit your user settings page. Simply scroll to the "Source Control Management" section to view any prior Bitbucket Server authorizations.
```
